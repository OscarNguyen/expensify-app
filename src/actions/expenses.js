import { v4 as uuidv4 } from 'uuid';
import database from '../firebase/firebase';

const exampleData = {
  description: '11pro',
  note: '',
  amount: 999,
  createdAt: 1234,
};
//add
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense,
});
export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const { description = '', note = '', amount = 0, createdAt = 0 } = expenseData;
    const expense = { description, note, amount, createdAt };

    return database
      .ref('expenses')
      .push(expense)
      .then((snapshot) => {
        dispatch(
          addExpense({
            id: snapshot.key,
            ...expense,
          }),
        );
      });
  };
};
//remove
export const removeExpense = ({ id }) => ({
  type: 'REMOVE_EXPENSE',
  id,
});

//edit
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates,
});

export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses,
});

export const startSetExpenses = () => {
  return (dispatch) => {
    return database
      .ref('expenses')
      .once('value')
      .then((dataSnapshot) => {
        console.log(dataSnapshot);
        const expenses = [];
        dataSnapshot.forEach((childSnapshot) => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val(),
          });
        });
        dispatch(setExpenses(expenses));
      });
  };
};
