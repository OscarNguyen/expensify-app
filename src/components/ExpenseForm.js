import React, { Component } from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

export default class ExpenseForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense.description && props.expense.description,
      note: props.expense.note && props.expense.note,
      amount: props.expense.amount && (props.expense.amount / 100).toString(),
      createdAt: props.expense.createdAt ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ''
    }
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState({
      description
    })
  }
  onNoteChange = e => {
    e.persist();
    this.setState({
      note: e.target.value
    })
  }
  onAmountChange = e => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState({ amount })
    }
  }

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState({
        createdAt
      })
    }

  }
  onFocusChange = ({ focused }) => {
    this.setState({
      calendarFocused: focused
    })
  }

  onSubmit = (e) => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState({ error: 'Please provide description and amount.' })
    } else {
      this.setState({ error: '' })
      this.props.onSubmit({
        description: this.state.description,
        note: this.state.note,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf()
      })
    }

  }
  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            value={this.state.description}
            onChange={this.onDescriptionChange}
            type='text'
            placeholder='Description'
            autoFocus
          />
          <input
            onChange={this.onAmountChange}
            value={this.state.amount}
            type='text'
            placeholder='Amount'
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            value={this.state.note}
            onChange={this.onNoteChange}
            placeholder='Add a note'
          >
          </textarea>
          <button>Add</button>
        </form>
      </div>
    )
  }
}