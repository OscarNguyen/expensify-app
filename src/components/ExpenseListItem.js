import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';
const ExpenseListItem = ({ id, description, amount, createdAt }) => (
  <div>
    <Link to={{ pathname: `/edit/${id}` }}>
      <h3>{description}</h3>
    </Link>
    <p>
      {numeral(amount / 100).format('$0,0.00')}-{moment(createdAt).format('MMMM Do, YYYY')}
    </p>
  </div>
);

/* const mapDispatchToProps = dispatch => {
  return {
    onClick: () => dispatch(removeExpense)
  }
} */
/* const mapStateToProps = (state) => {
  return {
    expenses: state.expenses
  }
} */
export default ExpenseListItem;
