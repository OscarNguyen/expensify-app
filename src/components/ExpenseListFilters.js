import React, { Component } from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByDate, sortByAmount, setEndDate, setStartDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilters extends Component {
  state = {
    calendarFocused: null,
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onTextChange = e => {
    this.props.setTextFilter(e.target.value);
  };

  onSortChange = e => {
    if (e.target.value === 'date') {
      this.props.sortByDate();
    } else if (e.target.value === 'amount') {
      this.props.sortByAmount();
    }
  };

  onFocusChange = calendarFocused => {
    this.setState({ calendarFocused });
  };

  render() {
    return (
      <div>
        <input value={this.props.filters.text} type="text" onChange={this.onTextChange} />
        <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
          <option value="amount">Amount</option>
          <option value="date">Date</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates={true}
        />
      </div>
    );
  }
}

const mapStateToProps = state => {
  return {
    filters: state.filters,
  };
};

const mapDispatchToProps = dispatch => ({
  setStartDate: date => dispatch.setStartDate(date),
  setEndDate: date => dispatch.setEndDate(date),
  setTextFilter: text => dispatch.setTextFilter(text),
  sortByDate: () => dispatch.sortByDate,
  sortByAmount: () => dispatch.sortByAmount,
});
export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(ExpenseListFilters);
