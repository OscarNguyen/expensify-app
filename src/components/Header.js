import React from 'react';
import { NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogout } from '../actions/auth';
export const Header = (props) => (
  <header>
    <h1>Expensify</h1>
    <NavLink to="/" exact={true} activeClassName="is-active">
      <p>Login Page</p>
    </NavLink>
    <NavLink to="/create" activeClassName="is-active">
      <p>Create Expense</p>
    </NavLink>

    <NavLink to="/help" activeClassName="is-active">
      <p>Help</p>
    </NavLink>
    <NavLink to="/dashboard" activeClassName="is-active">
      <p>Dashboard Page</p>
    </NavLink>
    <button onClick={props.startLogout}>Logout</button>
  </header>
);

const mapDispatchToProps = (dispatch) => ({
  startLogout: () => dispatch(startLogout()),
});
export default connect(undefined, mapDispatchToProps)(Header);
