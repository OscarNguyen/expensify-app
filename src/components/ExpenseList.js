import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = props => (
  <div>
    <h1>Expense List</h1>
    {props.expense.length === 0 ?
      (<p>No expenses</p>) : (
        props.expense.map(item => {
          return <ExpenseListItem key={item.id} {...item} />
        })
      )}
  </div>
);

const mapStateToProps = state => {
  return {
    expense: selectExpenses(state.expenses, state.filters)
  }
};

export default connect(mapStateToProps)(ExpenseList);
