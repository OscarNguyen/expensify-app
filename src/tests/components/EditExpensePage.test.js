import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';
//import { startRemoveExpense } from '../../actions/expenses';
let startEditExpense, startRemoveExpense, history, wrapper;

beforeEach(() => {
  startEditExpense = jest.fn();
  startRemoveExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <EditExpensePage
      startEditExpense={startEditExpense}
      history={history}
      startRemoveExpense={startRemoveExpense}
      expense={expenses[1]}
    />,
  );
});

test('should render EditExpensePage correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle startEditExpense', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);
  expect(history.push).toHaveBeenCalledWith('/dashboard');
  expect(startEditExpense).toHaveBeenCalledWith(expenses[1].id, expenses[1]);
});

test('should handle startRemoveExpense', () => {
  /* wrapper.find('button').prop('onClick')(); */
  wrapper.find('button').simulate('click');
  expect(history.push).toHaveBeenCalledWith('/dashboard');
  expect(startRemoveExpense).toHaveBeenCalledWith({ id: expenses[1].id });
});
