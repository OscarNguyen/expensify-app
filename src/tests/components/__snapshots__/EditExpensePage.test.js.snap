// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage correctly 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 6999,
        "createdAt": -345600000,
        "description": "iphone 11 pro max",
        "id": "2",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
